jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicRenderer.java

```diff
@@ -23,67 +21,74 @@
 
   private static final Logger log = LoggerFactory.getLogger(BasicRenderer.class);
-  protected Renderer.Vertex<V, E> vertexRenderer = new BasicVertexRenderer<V, E>();
-  protected Renderer.VertexLabel<V, E> vertexLabelRenderer = new BasicVertexLabelRenderer<V, E>();
-  protected Renderer.Edge<V, E> edgeRenderer = new BasicEdgeRenderer<V, E>();
-  protected Renderer.EdgeLabel<V, E> edgeLabelRenderer = new BasicEdgeLabelRenderer<V, E>();
+  protected Node<N, E> nodeRenderer = new BasicNodeRenderer<N, E>();
+  protected NodeLabel<N, E> nodeLabelRenderer = new BasicNodeLabelRenderer<N, E>();
+  protected Renderer.Edge<N, E> edgeRenderer = new BasicEdgeRenderer<N, E>();
+  protected Renderer.EdgeLabel<N, E> edgeLabelRenderer = new BasicEdgeLabelRenderer<N, E>();
 
   public void render(
-      RenderContext<V, E> renderContext,
-      VisualizationModel<V, E, Point2D> visualizationModel,
-      Spatial<V> spatial) {
-    if (spatial == null) {
+      RenderContext<N, E> renderContext,
+      VisualizationModel<N, E, Point2D> visualizationModel,
+      Spatial<N> nodeSpatial,
+      Spatial<E> edgeSpatial) {
+    if (nodeSpatial == null) {
       render(renderContext, visualizationModel);
       return;
     }
-    Set<V> visibleNodes = null;
+    Set<N> visibleNodes = null;
+    Set<E> visibleEdges = null;
+
     try {
       visibleNodes =
           (Set)
-              spatial.getVisibleNodes(
+              nodeSpatial.getVisibleElements(
                   ((VisualizationServer) renderContext.getScreenDevice()).viewOnLayout());
+
+      if (edgeSpatial != null) {
+        visibleEdges =
+            (Set)
+                edgeSpatial.getVisibleElements(
+                    ((VisualizationServer) renderContext.getScreenDevice()).viewOnLayout());
+      } else {
+        visibleEdges = visualizationModel.getNetwork().edges();
+      }
     } catch (ConcurrentModificationException ex) {
       // skip rendering until graph node index is stable,
       // this can happen if the layout relax thread is changing locations while the
       // visualization is rendering
+      log.info("got {} so returning", ex);
       return;
     }
 
     try {
-      Network<V, E> network = visualizationModel.getNetwork();
-      Set<E> visibleEdges = Sets.newHashSet(network.edges());
-      for (E edge : network.edges()) {
-        EndpointPair<V> endpoints = network.incidentNodes(edge);
-        boolean keep = false;
-        for (V v : endpoints) {
-          keep |= visibleNodes.contains(v);
-        }
-        if (!keep) {
-          visibleEdges.remove(edge);
-        }
-      }
+      Network<N, E> network = visualizationModel.getNetwork();
       // paint all the edges
+      log.trace("the visibleEdges are {}", visibleEdges);
       for (E e : visibleEdges) {
-        renderEdge(renderContext, visualizationModel, e);
-        renderEdgeLabel(renderContext, visualizationModel, e);
+        if (network.edges().contains(e)) {
```

Comment:
Are there circumstances under which network.edges() is not a superset of visibleEdges?  If not, this check seems redundant.  What led you to add it?

Reply:
Ah, makes sense.  Thanks.